### PR TITLE
debian: lock ceph user during purge

### PR DESCRIPTION
**Pull request by squidboylan**

### CODE REVIEW

File: debian/ceph-common.postrm [hunk 1]
```diff
@@ -26,6 +26,8 @@ case "$1" in
     purge)
 	rm -rf /var/log/ceph 
 	rm -rf /etc/ceph
+	userdel -f ceph
```
Comment (joscollin) [inline debian/ceph-common.postrm]: What about removing the files in the user's home directory along with the home directory itself using -r option ?
@liewegas Is it required to keep the files and home directory for future use ? Could you please suggest ?

I think the newer Debian Releases use deluser too.

**Comment by squidboylan** [comment]
Thanks for the input! http://tracker.ceph.com/issues/15872 is the issue this is trying to fix. If we cannot remove the ceph group and user, does it make sense to lock the user? I am worried that this would cause issues if someone uninstalls ceph, then installs it again, would this cause the installer to fail because it doesn't know how to unlock the user? I'm not sure what would happen in that case, but I would like to avoid causing issues in that situation. Thanks again!

**Comment by liewegas** [comment]
I think if you are in fact purging, and the package created the user, then you should remove it.

**Comment by tchaikov** [comment]
@squidboylan i think we should revise the postinst script to handle this case.

@liewegas but it's a security issue if we remove that user: uid reused.

**Comment by joscollin** [comment]
@squidboylan @tchaikov Considering the security as important, locking the user and updating postinst is better.

**Comment by squidboylan** [comment]
@joscollin can i get a review again?

File: debian/ceph-common.postinst [hunk 2]
```diff
@@ -62,6 +62,8 @@ case "$1" in
                -d $SERVER_HOME   \
                -g $SERVER_GROUP  \
                $SERVER_USER
+       # Unlock $SERVER_USER in case it is locked from an uninstall
+       usermod -e 99999 -U $SERVER_USER
```
Comment (tchaikov) [inline debian/ceph-common.postinst]: please drop the `-e 99999` option. and instead conditionalize this by `/etc/shadow`, like
```sh
    if [ -f /etc/shadow ]; then 
      usermod -U -e '' $SERVER_USER 
    else 
      usermod -U $SERVER_USER
    fi 
```

File: debian/ceph-common.postrm [hunk 3]
```diff
@@ -24,8 +24,9 @@ case "$1" in
     ;;
 
     purge)
-	rm -rf /var/log/ceph 
+	rm -rf /var/log/ceph
 	rm -rf /etc/ceph
+	usermod -L -e 1 ceph
```
Comment (tchaikov) [inline debian/ceph-common.postrm]: likewise, please put something like
```sh
    rm -rf 
    if [ -f /etc/shadow ]; then
      usermod -L -e 1 $SERVER_USER
    else
      usermod -L $SERVER_USER
    fi
```
and you need to query `/etc/default/ceph` for the username of the `ceph` user. it is might not be `ceph` if user overrides it.
```sh
[ -f "/etc/default/ceph" ] && . /etc/default/ceph
[ -z "$SERVER_USER" ] && SERVER_USER=ceph
```

File: debian/ceph-common.postinst (hunk 2 above)
Comment (joscollin) [inline debian/ceph-common.postinst]: @tchaikov Could you please explain why this just checks for the existence of the file `/etc/shadow` and not the existence of the locked user ?

**Comment by joscollin** [comment]
I think this is a topic for a debate. :-)

As @liewegas said, there is no point in leaving the user locked if we decided to purge and never going to install ceph again on that system.

On the other hand, as @tchaikov suggested, removing the user is a security risk too (uid reused).

**Comment by squidboylan** [comment]
I have made the changes requested. Has the discussion been had to decide what should be done on remove? If not how can that be moved forward so this can be merged?

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-08-19_02:32:27-rados-wip-kefu-testing-distro-basic-mira

**Pull request merged**